stm32cubemx: 6.0.1 -> 6.2.1

iFreilicht opened the pull request:
Upstream changes:
Main files moved from . to MX.
JAR file has no .exe extension anymore
Icon format changed from icns to ico.
There is now a bundled JRE, should we use it? It may only be compatible with Linux, not macOS.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Update to a never version of CubeMX. All changes since 6.0.1 can be found [here (PDF)](https://www.st.com/content/ccc/resource/technical/document/release_note/b2/02/4b/e3/81/32/49/c4/DM00107607.pdf/files/DM00107607.pdf/jcr:content/translations/en.DM00107607.pdf).

Unfortunately, it seems like CubeMX doesn't work with the newer OpenJDK16, which contains the JRE on master. OpenJDK 15 works, though. 
The two derivations I tested against:

- 9ryn07zmn2rp9nsiyvpq0jcw771lrgmi-openjdk-15.0.1-ga
- b8pk4qfj00rppf3nn0jzqi5va5d543mp-openjdk-16+36

**I'm not quite sure how to proceed with testing here.**

@wucke13

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

raboof commented:
> There is now a bundled JRE, should we use it?

I think we should strongly prefer the nixpkgs-packaged JRE. (in particular cases, if upstream feels strongly we should be using their version (or even have patched it), we could consider packaging theirs - that should be rare exceptions though IMHO).

> it seems like CubeMX doesn't work with the newer OpenJDK16, which contains the JRE on master. OpenJDK 15 works, though.

So 6.0.1 worked with jdk16 but 6.2.1 doesn't? I'd probably report this issue upstream, and first try if falling back to the last LTS (jdk11) ork. If it doesn't I guess we could consider `adoptopenjdk-hotspot-bin-15`.

wucke13 commented:
I also strongly oppose using the bundled JRE. I got to admit that I'm kind of surprised that it broke on the newest JRE at all, I didn't have the impression that cubemx does anything wild. Using an LTS version is probably a "good enough" solution.

r-rmcgibbo commented:
Result of `nixpkgs-review pr 120341` at 91c75455 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>stm32cubemx</li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: unused-argument

Unused argument: `requireFile`.
Near pkgs/development/tools/misc/stm32cubemx/default.nix:1:16:
```
  |
1 | { lib, stdenv, requireFile, makeDesktopItem, icoutils, fdupes, imagemagick, jre, fetchzip }:
  |                ^
```
</li>
  </ul>
</details>

SuperSandro2000 commented on pkgs/development/tools/misc/stm32cubemx/default.nix:
```diff
@@ -12,36 +12,37 @@ let
 in
 stdenv.mkDerivation rec {
   pname = "stm32cubemx";
+  iconame = "STM32CubeMX";
```
Please remove this variable. I don't think it is really required.

iFreilicht replied:
Unfortunately, it is. With this new version, the filename and format of the desktop icon was changed. See lines 36, 42 and 45

SuperSandro2000 replied:
We should hardcode it there. Right now it is also exportet as an env. Alternative would be to move it into a  let in.

iFreilicht replied:
Ah, good point! I moved it to the let in :)

iFreilicht commented:
> So 6.0.1 worked with jdk16 but 6.2.1 doesn't? I'd probably report this issue upstream, and first try if falling back to the last LTS (jdk11) ork. If it doesn't I guess we could consider `adoptopenjdk-hotspot-bin-15`.

No, 6.0.1 didn’t even try to use jdk16 when I installed it with `nix-env -iA nixpkgs.stm32cubemx`. I think the reason might be that the nixpkgs channel im my `.nix-defexpr` was still outdated. When I used `nix-env -f ~/repos/nixpkgs -iA stm32cubemx` to install my updated version, it also installed jdk16.

I’ll test your suggestions tomorrow and will report this upstream. I could imagine that their response will be to just use the bundled JRE, though.

> I also strongly oppose using the bundled JRE. I got to admit that I'm kind of surprised that it broke on the newest JRE at all, I didn't have the impression that cubemx does anything wild. Using an LTS version is probably a "good enough" solution.

Yeah, that’s weird. It’s not something inside the UI, it happens immediately on startup.

iFreilicht commented:
So, even the old version 6.0.1 does NOT work with JDK16, so this package could actually be considered broken right now. The only reason it installed properly for me was that my nixpkgs channel wasn't updated yet.

I tested this on a different machine and could confirm this:

<details>
  <summary>Click to expand!</summary>

```bash
# nixpkgs wasn't updated yet
$ NIXPKGS_ALLOW_UNFREE=1 nix-env -iA nixpkgs.stm32cubemx
[...]
mkdir: created directory '/nix/store/b09nvyxa68n3f8zsqbn2l9s9gqgmksq6-stm32cubemx-6.0.1/share/icons'
[...]
$ nix-store -q  --references /nix/store/b09nvyxa68n3f8zsqbn2l9s9gqgmksq6-stm32cubemx-6.0.1
/nix/store/cwnwyy82wrq53820z6yg7869z8dl5s7g-bash-4.4-p23
/nix/store/v8dykc8xwq8f0gv3pcca4df8j8kvqsga-openjdk-15.0.1-ga
/nix/store/xir9arx16wlig1v0wikp31gwwjy6l0n2-stm32CubeMX.desktop
/nix/store/b09nvyxa68n3f8zsqbn2l9s9gqgmksq6-stm32cubemx-6.0.1
$ /nix/store/b09nvyxa68n3f8zsqbn2l9s9gqgmksq6-stm32cubemx-6.0.1/bin/stm32cubemx
[...] # lots of output, works perfectly
$ nix-channel --list
nixpkgs https://nixos.org/channels/nixpkgs-unstable
$ nix-channel --update
unpacking channels...
warning: Nix search path entry '/nix/var/nix/profiles/per-user/root/channels' does not exist, ignoring
created 1 symlinks in user environment
$ NIXPKGS_ALLOW_UNFREE=1 nix-env -iA nixpkgs.stm32cubemx
[...]
mkdir: created directory '/nix/store/n19vahbsrfhgmbj4kx1ws98npmjw6wsv-stm32cubemx-6.0.1/share/icons'
[...]
created 45 symlinks in user environment
$ nix-store -q  --references /nix/store/b09nvyxa68n3f8zsqbn2l9s9gqgmksq6-stm32cubemx-6.0.1
/nix/store/b8pk4qfj00rppf3nn0jzqi5va5d543mp-openjdk-16+36
/nix/store/bhqm8m0svacc4ps1hbgqhqa3iz4g3256-stm32CubeMX.desktop
/nix/store/jdi2v7ir1sr6vp7pc5x0nhb6lpcmg6xg-bash-4.4-p23
/nix/store/n19vahbsrfhgmbj4kx1ws98npmjw6wsv-stm32cubemx-6.0.1
$ /nix/store/n19vahbsrfhgmbj4kx1ws98npmjw6wsv-stm32cubemx-6.0.1/bin/stm32cubemx
[...] # lots of output, and the splashscreen appears, but:
Exception in thread "AWT-EventQueue-0" java.lang.ExceptionInInitializerError
	at com.thoughtworks.xstream.XStream.setupConverters(XStream.java:990)
	at com.thoughtworks.xstream.XStream.<init>(XStream.java:593)
	at com.thoughtworks.xstream.XStream.<init>(XStream.java:515)
	at com.thoughtworks.xstream.XStream.<init>(XStream.java:484)
	at com.thoughtworks.xstream.XStream.<init>(XStream.java:430)
	at com.thoughtworks.xstream.XStream.<init>(XStream.java:397)
	at com.st.library.rawvaluescontainer.FilesUtils.<clinit>(FilesUtils.java:41)
	at com.st.library.analyticscore.AnalyticsContext.<init>(AnalyticsContext.java:167)
	at com.st.library.dynamicproxy.AnalyticsInvocationHandler.<init>(AnalyticsInvocationHandler.java:36)
	at com.st.library.builder.AnalyticsBuilder.getAnalytics(AnalyticsBuilder.java:33)
	at com.st.library.analyticscore.AnalyticsAPI.initialize(AnalyticsAPI.java:132)
	at com.st.microxplorer.plugins.analytics.AnalyticsPlugin.<init>(AnalyticsPlugin.java:165)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:78)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
	at java.base/java.lang.reflect.ReflectAccess.newInstance(ReflectAccess.java:128)
	at java.base/jdk.internal.reflect.ReflectionFactory.newInstance(ReflectionFactory.java:350)
	at java.base/java.lang.Class.newInstance(Class.java:642)
	at com.st.microxplorer.plugin.PluginManage.loadPluginJar(PluginManage.java:335)
	at com.st.microxplorer.plugin.PluginManage.PluginSearch(PluginManage.java:206)
	at com.st.microxplorer.plugin.PluginManage.PluginDiscovery(PluginManage.java:49)
	at com.st.microxplorer.maingui.MainPanel.initialization(MainPanel.java:236)
	at com.st.microxplorer.maingui.MainFrame.initialization(MainFrame.java:149)
	at com.st.microxplorer.maingui.MicroXplorer.run(MicroXplorer.java:226)
	at com.st.microxplorer.maingui.STM32CubeMX.run(STM32CubeMX.java:25)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:316)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:770)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:715)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:391)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:740)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
Caused by: java.lang.reflect.InaccessibleObjectException: Unable to make field private final java.util.Comparator java.util.TreeMap.comparator accessible: module java.base does not "opens java.util" to unnamed module @13663f96
	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:357)
	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:297)
	at java.base/java.lang.reflect.Field.checkCanSetAccessible(Field.java:177)
	at java.base/java.lang.reflect.Field.setAccessible(Field.java:171)
	at com.thoughtworks.xstream.core.util.Fields.locate(Fields.java:40)
	at com.thoughtworks.xstream.converters.collections.TreeMapConverter.<clinit>(TreeMapConverter.java:50)
	... 39 more
```

</details>

No idea what the actual error means, but it's probably not fixable on our end.

I tried it with jdk11 as suggested, and that seems to work perfectly 🎉

I also sent [a bug report to ST](https://community.st.com/s/question/0D53W00000jnOzPSAU/stm32cubemx-crashes-on-launch-with-openjdk16).

> Result of `nixpkgs-review pr 120341` at [91c7545](https://github.com/NixOS/nixpkgs/commit/91c75455a89a82c887e9804b7d28637d8a11720e) run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
> 1 package built successfully:
> 1 suggestion:
> 
>     * warning: unused-argument
>       Unused argument: `requireFile`.
>       Near pkgs/development/tools/misc/stm32cubemx/default.nix:1:16:
>       ```
>         |
>       1 | { lib, stdenv, requireFile, makeDesktopItem, icoutils, fdupes, imagemagick, jre, fetchzip }:
>         |                ^
>       ```

I applied this change as well, works perfectly still.

iFreilicht commented:
Ok, I'd say this is pretty much mergable now. I checked the size impact, and this update actually reduces the closure size by ~150MB:

```
$ nix path-info -S /nix/store/n19vahbsrfhgmbj4kx1ws98npmjw6wsv-stm32cubemx-6.0.1
/nix/store/n19vahbsrfhgmbj4kx1ws98npmjw6wsv-stm32cubemx-6.0.1      1918445376
$ nix path-info -S /nix/store/r6mjqy87mlvna6xqyccm3qbajv8i6djl-stm32cubemx-6.2.1/ 
/nix/store/r6mjqy87mlvna6xqyccm3qbajv8i6djl-stm32cubemx-6.2.1      1765157296
```

Should I squash the commits or will you merge them as-is?

raboof commented:
> Caused by: java.lang.reflect.InaccessibleObjectException: Unable to make field private final java.util.Comparator java.util.TreeMap.comparator accessible: module java.base does not "opens java.util" to unnamed module @13663f96

This is due to the fact that JDK16 has become more strict about reflective access across modules. Thanks for reporting upstream, and good find that it looks like it's already fixed in the newer version of xstream.

> I tried it with jdk11 as suggested, and that seems to work perfectly :tada: 

Nice!

> Should I squash the commits or will you merge them as-is?

If you could squash that'd be great!

raboof commented on pkgs/development/tools/misc/stm32cubemx/default.nix:
```diff
@@ -1,47 +1,50 @@
-
+{ lib, stdenv, makeDesktopItem, icoutils, fdupes, imagemagick, jdk11, fetchzip }:
+# TODO: JDK16 causes STM32CubeMX to crash right now, so we fixed the version to JDK11
+# This may be fixed in a future version of STM32CubeMX. If you're updating this derivation,
```
Maybe add a link to the upstream issue

iFreilicht replied:
Done!

raboof reviewed (changes requested):
(LGTM, but please squash before merging)

iFreilicht commented:
@raboof Commit is squashed and I tested it once more. Seems to work perfectly!

SuperSandro2000 commented on pkgs/development/tools/misc/stm32cubemx/default.nix:
```diff
@@ -1,47 +1,52 @@
 let
-  version = "6.0.1";
+  version = "6.2.1";
   desktopItem = makeDesktopItem {
     name = "stm32CubeMX";
     exec = "stm32cubemx";
     desktopName = "STM32CubeMX";
     categories = "Development;";
     icon = "stm32cubemx";
   };
```
This can be moved into a copyDesktopItems

iFreilicht replied:
I'll have a look, thanks!

iFreilicht replied:
Done!

SuperSandro2000 commented on pkgs/development/tools/misc/stm32cubemx/default.nix:
```diff
@@ -1,47 +1,52 @@
+# and try replacing all occurrences of jdk11 with jre and test whether it works.
 let
-  version = "6.0.1";
+  version = "6.2.1";
```
```suggestion
```

SuperSandro2000 commented on pkgs/development/tools/misc/stm32cubemx/default.nix:
```diff
@@ -1,47 +1,52 @@
 in
 stdenv.mkDerivation rec {
   pname = "stm32cubemx";
   inherit version;
 
 
   src = fetchzip {
```
```suggestion
  version = "6.2.1";

  src = fetchzip {
```

iFreilicht replied:
Can fetchzip still work in that case? It needs `version` to construct the download link.

iFreilicht replied:
Also, I don't want to be pedantic, but do we really need to change much more about this package? Is it good practice to have as little `let` or `inherit` statements as possible?

iFreilicht replied:
Ah right, it works because of `rec`. Done!

iFreilicht committed:
stm32cubemx: 6.0.1 -> 6.2.1

Fixes:
STM32CubeMX uses an outdated version of xstream that is not
compatible with JDK16. The derivation uses JDK11 (LTS) explicitly now.
Additionally, the desktop file wasn't generated by copyDesktopItems
before. This is now fixed as well.

Upstream changes:
- Main files moved from . to MX
- JAR file has no .exe extension anymore
- Icon format changed from icns to ico
- There is now a bundled JRE, but we prefer not to use it

Additional changes:
- Move version definition into mkDerivation

iFreilicht commented:
@SuperSandro2000 All your suggestions were applied, please have another look :)
